src/core/public/application/types.ts

```diff
@@ -245,6 +266,13 @@ export interface App<HistoryLocationState = unknown> {
    * ```
    */
   exactRoute?: boolean;
+
+  /**
+   * The visibility of the application based on workspace.
+   * Defaulting to `both`
+   * See {@link AppVisibility}
+   */
+  visibility?: AppVisibility;
```

Comment:
There should be 3 states: both, homeOnly and workspaceOnly. I have some concern that regarding `undefined` as both would introduce some confusion.

Reply:
Though actually a link should be either homeOnly or workspaceOnly, I am afraid we can not force plugins to change the registration by 2.14 .

Reply:
Just a little concerned about there are too many visibility related property (status, navLinkStatus, chromeless) in the interface.  Or we can just renaming to `workspaceLess`, then we just mark home plugin to true.

Reply:
Yes, agree on that there are too many similar property. So I think maybe we could extend the status field by adding `inaccessible-in-workspace` and `inaccessible-in-home`, what do you think?

Reply:
+1 on `workspaceless`, similar with `chromeless`. applicatioin with `workspaceless` as true will not have workspace as its context.

Reply:
The status property has been used in many places. We need to update all the inaccessible cases if add two more inaccessible statuses. Add a new field is make sense to me, just concerned about the property name if these logic are only for workspace.

Reply:
Forget that, I do not think coupling workspace related status into status field is a good idea.

Reply:
Agree, maybe we could use some boolean value like `workspaceless`

Reply:
Thanks for the thought and comment, will summarize those in the official issue. 

And we are gonna to use `workspaceless` to indicate the app is not accessible within a workspace.

Reply:
`workspaceless: true` -> the app is inaccessible within a workspace
`workspaceless: false` -> the app can be accessed both in a workspace or outside of a workspace

There is no state to represent **an app can only be accessed within a workspace**, but this is valid use case, apps such as `visualization`, `dashboards`, etc are not going to be accessible outside of a workspace in the future based on the design.

Reply:
Right, it reminds me that we need a flag to know which nav is visible when in home workspace. In current design, it seems a boolean value is not enough for such case.

Reply:
Yes to "apps such as visualization, dashboards, etc are not going to be accessible outside of a workspace in the future based on the design". And Hailong's examples are also valid. One more example is the workspace list page.

Reply:
> workspace update/overview page are two examples

As Suzhou suggested, only register workspace overview/update app when in a workspace. so these two can be avoid in another(better) way.

Reply:
> > an app can only be accessed within a workspace
> 
> workspace update/overview page are two examples

I mean only two state is not enough, if using `workspaceless` which is boolean, it can only represent two state above.

Reply:
Is it necessary to mark the application as "an app can only be accessed within a workspace"? It will a little bit confused when workspace not enabled.

Reply:
> I mean only two state is not enough

I see. do you think undefined/null, true, false will be three states and enough for our case? `undefined/null` same as current.

Reply:
If we need to represent 3 states in a single field, an enum would be better instead of regarding `undefined/null` as a different state to `false`. This is a draft PR and we can adjust that.
What about:
- `workspaceAccessibility.withinWorkspace` -> only accessible in workspace, should be filter out from home in the future when workspace enabled, which means user won't be enable to create visualizations and dashboards out of workspace.
- `workspaceAccessibility.outOfWorkspace` -> only accessible out of workspace, can be used for pages like home/workspace list/workspace create page,
- `workspaceAccessibility.both` | undefined -> as current to keep backward capability.

Reply:
> Is it necessary to mark the application as "an app can only be accessed within a workspace"? It will a little bit confused when workspace not enabled. For these pages, I think we still need to be displayed when workspace is turn off.

That's a good point. One way that I can think of is to check the workspace flag when plugin register the app, for example, via capabilities service. If workspace is turned off, plugin should not use this flag